CPP-4982,CPP-4913 S871 Mention that copies are allowed. (#3659)
SonarSource/rspec: rules/S871/cfamily/rule.adoc

=== Exceptions

Void casts are allowed:
[source,cpp]
----
(void)call(); // Compliant
void(call()); // Compliant
----

Furthermore functional cast can be used, in case when it lead to copy or constructor call:
[source,cpp]
----
class Clazz {
  Clazz(int, int);
};
Clazz create();

template<typename T>
void process(T&& t);

void foo(int i, Clazz c) {
  process(Clazz(1, 2));     // Compliant, consructor call

>   process(Clazz(c));        // Compliant, copy

Author: ```suggestion
  process(auto(c));        // Compliant, copy
```

=== Exceptions

> Void casts are allowed:

Reviewer: ```suggestion
This rule ignores void casts. They are a common pattern to express that an expression is dropped on purpose.

```

[source,cpp]
----

> (void)call(); // Compliant
> void(call()); // Compliant

Reviewer: ```suggestion
(void)call(); // Compliant by exception
void(call()); // Compliant by exception
```

----

> Furthermore functional cast can be used, in case when it lead to copy or constructor call:

Reviewer: ```suggestion
Furthermore, this rule ignores functional casts when they lead to calls to constructors, because their purpose is clear and no dangerous conversion can happen.

```

[source,cpp]
----
class Clazz {
  Clazz(int, int);
};
Clazz create();

template<typename T>
void process(T&& t);

void foo(int i, Clazz c) {

>   process(Clazz(1, 2));     // Compliant, consructor call
>   process(Clazz(c));        // Compliant, copy
>   process(auto(c));         // Compliant, copy
>   process(Clazz(create())); // Compliant, copy
>   process((Clazz)c);        // Noncomplaint, C-Style cast
>  
>   process(int(i));  // Compliant, copy
>   process(auto(i)); // Compliant, copy
>   process((int)i);  // Noncompliant, C-style cast

Reviewer: ```suggestion
  process(Clazz(1, 2));     // Compliant: constructor call
  process(Clazz(c));        // Compliant: copy
  process(auto(c));         // Compliant: copy
  process(Clazz(create())); // Compliant: copy
  process((Clazz)c);        // Noncompliant: C-Style cast
 
  process(int(i));  // Compliant: copy
  process(auto(i)); // Compliant: copy
  process((int)i);  // Noncompliant: C-style cast
```

=== Exceptions

This rule ignores void casts. They are a common pattern to express that an expression is dropped on purpose.

[source,cpp]
----
(void)call(); // Compliant by exception
void(call()); // Compliant by exception
----

> Furthermore, this rule ignores functional casts when they lead to calls to constructors, because their purpose is clear and no dangerous conversion can happen.

Author: `int(10)` is not consructor, and `Class(create())` also not. I want to capture that, what about?
```suggestion
Furthermore, this rule ignores functional casts when creating an object of the same type (copy) of the object or leads to calls to constructors, because their purpose is clear and no dangerous conversion can happen.
```

> Furthermore, this rule ignores functional casts when creating an object of the same type (copy) of the object or leads to calls to constructors, because their purpose is clear and no dangerous conversion can happen.

Reviewer: I would suggest this:
```suggestion
Furthermore, this rule ignores functional casts when creating an object of the same type (copy) or when they lead to calls to a constructor. In those cases, the purpose is clear and no dangerous conversion can happen.
```
